### PR TITLE
UCP/GTEST: Fix and test ucp requests leak from the ptr_map

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

Fix and test ucp requests leak from the ptr_map.

## Why ?

To prevent having leaked requests in ptr map.

## How ?

1. Save request ID to the request to be able remove them just getting an ID form request.
2. Delete IDs where it is needed.
3. Fix AMO/RMA protocols to generate request ID from progress function (release an allocated ID if send was not OK) rather than from pack_cb that could be invoked several times.

### CODE REVIEW

**Comment by dmitrygx** [comment]
@alinask @brminich @hoopoepg could you review pls?

File: src/ucp/core/ucp_request.h [hunk 1]
```diff
@@ -237,13 +245,14 @@ struct ucp_request {
...
                 struct {
+                    ucs_ptr_map_key_t     sreq_id;     /* Send request ID */
                     uint64_t              remote_addr; /* Remote address */
                     ucp_rkey_h            rkey;        /* Remote memory key */
                     uint64_t              value;       /* Atomic argument */
-                    uct_atomic_op_t       uct_op;      /* Requested UCT AMO */
                 } amo;
 
                 struct {
+                    ucs_ptr_map_key_t sreq_id;   /* Send request ID */
```
Comment (alinask) [inline src/ucp/core/ucp_request.h:255]: why do we need this in offload if it doesn't have error handling support?

File: src/ucp/core/ucp_request.c [hunk 2]
```diff
@@ -338,6 +338,14 @@ void ucp_request_init_multi_proto(ucp_request_t *req,
     UCS_PROFILE_REQUEST_EVENT(req, multi_func_str, req->send.length);
 }
 
+static UCS_F_ALWAYS_INLINE void ucp_send_request_init_id(ucp_request_t *req)
+{
+    if (req->flags & UCP_REQUEST_FLAG_SYNC) {
+        ucs_assert(req->flags & UCP_REQUEST_FLAG_SEND_TAG);
+        req->send.msg_proto.sreq_id = ucp_send_request_get_id(req);
```
Comment (yosefe) [inline src/ucp/core/ucp_request.c:345]: common inline helper with ucp_am.c:763 and tag_rndv.c:108 ?

File: src/ucp/core/ucp_request.h [hunk 3]
```diff
@@ -204,7 +212,7 @@ struct ucp_request {
                 } rkey_ptr;
 
                 struct {
-                    ucs_ptr_map_key_t req_id;         /* the send request ID on receiver side */
+                    ucs_ptr_map_key_t sreq_id;        /* the send request ID on receiver side */
```
Comment (yosefe) [inline src/ucp/core/ucp_request.h:215]: should it be rreq_id rhen? see the comment

File: src/ucp/tag/eager_rcv.c [hunk 4]
```diff
@@ -259,6 +259,8 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_eager_offload_sync_ack_handler,
     ucs_queue_for_each_safe(sreq, iter, queue, send.tag_offload.queue) {
         if ((sreq->send.tag_offload.ssend_tag == rep_hdr->sender_tag) &&
             (ucp_ep_local_id(sreq->send.ep) == rep_hdr->ep_id)) {
+            ucp_worker_del_request_id(worker, sreq,
+                                      sreq->send.tag_offload.sreq_id);
```
Comment (yosefe) [inline src/ucp/tag/eager_rcv.c:263]: is it really needed here - can replace by assert?

File: src/ucp/tag/offload.c [hunk 5]
```diff
@@ -565,6 +565,10 @@ static void ucp_tag_offload_rndv_zcopy_completion(uct_completion_t *self)
 {
     ucp_request_t *req = ucs_container_of(self, ucp_request_t,
                                           send.state.uct_comp);
+
+    ucs_assert(!ucp_ep_use_indirect_id(req->send.ep));
+    ucp_worker_del_request_id(req->send.ep->worker, req,
+                              req->send.tag_offload.sreq_id);
```
Comment (yosefe) [inline src/ucp/tag/offload.c:571]: is it really needed?

File: src/ucs/datastruct/ptr_map.inl [hunk 6]
```diff
@@ -79,6 +79,7 @@ ucs_ptr_map_put(ucs_ptr_map_t *map, void *ptr, int indirect,
     if (ucs_likely(!indirect)) {
         *key = (uintptr_t)ptr;
         ucs_assert(!(*key & UCS_PTR_MAP_KEY_MIN_ALIGN));
+        ucs_assert(*key != 0);
```
Comment (yosefe) [inline src/ucs/datastruct/ptr_map.inl:82]: why needed?
maybe assert in func start that ptr != NULL?

File: src/ucs/datastruct/ptr_map.inl [hunk 7]
```diff
@@ -131,6 +132,7 @@ ucs_ptr_map_extract(ucs_ptr_map_t *map, ucs_ptr_map_key_t key)
     void *value;
 
     if (ucs_likely(!(key & UCS_PTR_MAP_KEY_INDIRECT_FLAG))) {
+        ucs_assert(key != 0);
```
Comment (yosefe) [inline src/ucs/datastruct/ptr_map.inl:135]: why needed?

File: src/ucp/rma/rma_sw.c [hunk 8]
```diff
@@ -72,14 +72,19 @@ static ucs_status_t ucp_rma_sw_progress_get(uct_pending_req_t *self)
...
-    req->send.lane = ucp_ep_get_am_lane(req->send.ep);
-    status         = ucp_rma_sw_do_am_bcopy(req, UCP_AM_ID_GET_REQ,
-                                            req->send.lane,
-                                            ucp_rma_sw_get_req_pack_cb, req,
-                                            &packed_len);
-    if ((status != UCS_OK) && (status != UCS_ERR_NO_RESOURCE)) {
-        /* completed with error */
-        ucp_request_complete_send(req, status);
+    req->send.lane        = ucp_ep_get_am_lane(req->send.ep);
+    req->send.rma.sreq_id = ucp_send_request_get_id(req);
```
Comment (brminich) [inline src/ucp/rma/rma_sw.c:76]: why not adding in pack cb?
or add if operations completes ok?

File: src/ucs/datastruct/ptr_map.inl (hunk 6 above)
Comment (alinask) [reply src/ucs/datastruct/ptr_map.inl:82]: @yosefe due to the discussion here - https://github.com/openucx/ucx/pull/5769/files#r503383561

File: src/ucs/datastruct/ptr_map.inl (hunk 7 above)
Comment (alinask) [reply src/ucs/datastruct/ptr_map.inl:135]: @yosefe due to the discussion here - https://github.com/openucx/ucx/pull/5769/files#r503383561

File: src/ucp/core/ucp_request.h (hunk 3 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_request.h]: better to name it as `remote_req_id`

File: src/ucp/core/ucp_request.h (hunk 1 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_request.h]: removed

File: src/ucp/core/ucp_request.h [hunk 9]
```diff
@@ -273,7 +282,10 @@ struct ucp_request {
                 ucp_lane_index_t  am_bw_index;     /* AM BW lane index */
                 ucp_lane_map_t    lanes_map_avail; /* Used lanes map */
             };
-            uint8_t               mem_type;        /* Memory type */
+            uint8_t               uct_atomic_op;   /* Requested UCT AMO, values are
```
[the remark this answers is not in the document]
Comment (dmitrygx) [reply src/ucp/core/ucp_request.h]: retuned back per our offline discussion

File: src/ucp/core/ucp_worker.inl [hunk 10]
```diff
@@ -66,6 +66,9 @@ ucp_worker_get_request_id(ucp_worker_h worker, ucp_request_t *req, int indirect)
     ucs_ptr_map_key_t id;
     ucs_status_t status;
 
+    ucs_assert(!(req->flags & UCP_REQUEST_FLAG_IN_PTR_MAP));
+    req->flags |= UCP_REQUEST_FLAG_IN_PTR_MAP;
```
[the remark this answers is not in the document]
Comment (dmitrygx) [reply src/ucp/core/ucp_worker.inl]: done

File: src/ucp/core/ucp_worker.inl [hunk 11]
```diff
@@ -86,10 +89,13 @@ ucp_worker_get_request_by_id(ucp_worker_h worker, ucs_ptr_map_key_t id)
...
 
 static UCS_F_ALWAYS_INLINE void
-ucp_worker_del_request_id(ucp_worker_h worker, ucs_ptr_map_key_t id)
+ucp_worker_del_request_id(ucp_worker_h worker, ucp_request_t *request,
+                          ucs_ptr_map_key_t id)
 {
     ucs_status_t status UCS_V_UNUSED;
 
+    ucs_assert(request->flags & UCP_REQUEST_FLAG_IN_PTR_MAP);
+    request->flags &= ~UCP_REQUEST_FLAG_IN_PTR_MAP;
```
[the remark this answers is not in the document]
Comment (dmitrygx) [reply src/ucp/core/ucp_worker.inl]: done

File: src/ucp/rma/amo_basic.c [hunk 12]
```diff
@@ -56,12 +56,12 @@ static ucs_status_t ucp_amo_basic_progress_post(uct_pending_req_t *self)
...
     ucp_request_t *req    = ucs_container_of(self, ucp_request_t, send.uct);
-    ucp_rkey_h rkey       = req->send.amo.rkey;
     ucp_ep_t *ep          = req->send.ep;
-    uint64_t value        = req->send.amo.value;
     uint64_t *result      = req->send.buffer;
+    uct_atomic_op_t op    = req->send.uct_atomic_op;
+    uint64_t value        = req->send.amo.value;
     uint64_t remote_addr  = req->send.amo.remote_addr;
-    uct_atomic_op_t op    = req->send.amo.uct_op;
+    ucp_rkey_h rkey       = req->send.amo.rkey;
```
[the remark this answers is not in the document]
Comment (dmitrygx) [reply src/ucp/rma/amo_basic.c]: fixed

File: src/ucp/rma/rma_sw.c (hunk 8 above)
Comment (dmitrygx) [reply src/ucp/rma/rma_sw.c:76]: inside the pack_cb we don't actually know if AM Bcopy completes with success or not.
so, we do it prior to calling AM Bcopy in order to makes the sreq_id initialized for further usage

File: src/ucp/tag/eager_rcv.c (hunk 4 above)
Comment (dmitrygx) [reply src/ucp/tag/eager_rcv.c]: replaced by assert

File: src/ucp/tag/offload.c (hunk 5 above)
Comment (dmitrygx) [reply src/ucp/tag/offload.c]: replaced by assert

File: src/ucp/core/ucp_worker.inl [hunk 13]
```diff
@@ -96,7 +96,8 @@ ucp_worker_del_request_id(ucp_worker_h worker, ucp_request_t *request,
 {
     ucs_status_t status UCS_V_UNUSED;
 
-    ucs_assert(request->flags & UCP_REQUEST_FLAG_IN_PTR_MAP);
+    ucs_assert((request->flags & UCP_REQUEST_FLAG_IN_PTR_MAP) ||
+               !ucs_ptr_map_key_indirect(id));
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.inl:100]: maybe add helper for this assert? seems to be in many places

File: src/ucs/datastruct/ptr_map.inl [hunk 14]
```diff
@@ -53,6 +53,18 @@ static inline void ucs_ptr_map_destroy(ucs_ptr_map_t *map)
...
 }
 
+/**
+ * Returns whether the key is indirect or not.
+ *
+ * @param [in]  key     Key to object pointer.
+ *
+ * @return 0 - direct, otherside - indirect.
+ */
+static inline int ucs_ptr_map_key_indirect(ucs_ptr_map_key_t key)
```
Comment (yosefe) [inline src/ucs/datastruct/ptr_map.inl:63]: UCS_F_ALWAYS_INLINE

File: src/ucp/core/ucp_request.c (hunk 2 above)
Comment (yosefe) [reply src/ucp/core/ucp_request.c:345]: @dmitrygx WDYT?

File: src/ucs/datastruct/ptr_map.inl (hunk 14 above)
Comment (dmitrygx) [reply src/ucs/datastruct/ptr_map.inl:63]: other functions have `inline`
is it ok to declare this function with `UCS_F_ALWAYS_INLINE`?

File: src/ucp/core/ucp_worker.inl (hunk 13 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_worker.inl:100]: makes sense, but need two different helper functions

File: src/ucs/datastruct/ptr_map.inl (hunk 14 above)
Comment (dmitrygx) [reply src/ucs/datastruct/ptr_map.inl:63]: never mind, I was wrong
fixed

**Review by yosefe — APPROVED** [review]

**Review by brminich — APPROVED** [review]

**Comment by dmitrygx** [comment]
@brminich @yosefe @alinask thank you for the review
changes are squashed

@alinask do you have additional comments?

**Comment by alinask** [comment]
> @brminich @yosefe @alinask thank you for the review
> changes are squashed
> 
> @alinask do you have additional comments?

@dmitrygx I don't have additional comments.

Only one short request.
Could you please run io_demo with this PR with an Eth IP (not RoCE)? to make sure there isn't a leak.
Running io_demo was what releveled this problem in the first place and it's not part of CI yet (with Eth).
One server and one client on different hosts is enough. the original command line was:
server: `UCX_SOCKADDR_CM_ENABLE=y ./io_demo -o write,read -d 128:4194304 -i 100000 -w 10`
client: ` UCX_SOCKADDR_CM_ENABLE=y ./io_demo -o write,read -d 128:4194304 -i 100000 -w 10 <ETH_IP>`

Thanks!

**Comment by dmitrygx** [comment]
> Could you please run io_demo with this PR with an Eth IP (not RoCE)? to make sure there isn't a leak.

@alinask sure, thank you for giving me a reproducer.
I ran the test as you requested. It works fine when I specified IP over commodity Ethernet, no leaks were revealed during the test.
I tested it *with* and *without* `UCX_TLS="tcp,sockcm"`, because the master branch selects IB (if it exists) even if we specified an IP address of an Ethernet device (establishes a connection thru sockcm and TCP, and then switches to more optimal transport).
```
[1607977721.761084] [UCX-connection #1 10.210.0.125:1337] destroying, ep is 0x7fec8111d000
[1607977721.761133] [UCX-connection #1 10.210.0.125:1337] closing ep 0x7fec8111d000 mode force
[1607977721.761254] [UCX-connection #1 10.210.0.125:1337] released
[1607977721.761260] [DEMO] Client exit with status 'OK'
[1607977721.761263] [UCX] destroy_connections
[1607977721.761505] [UCX] destroy_connections
[1607977721.761519] [UCX] iomsg receive request 0xa2ba80 failed: Request canceled
```

Running the test on the master branch completes with the warning at the end of the testing when `UCX_TLS="tcp,sockcm"` is specified (but if an IB transport is used, it completes w/o any warnings):
```
...
[1607977457.603803] [UCX-connection #1 10.210.0.125:1337] destroying, ep is 0x7f8aa1787000
[1607977457.603821] [UCX-connection #1 10.210.0.125:1337] closing ep 0x7f8aa1787000 mode force
[1607977457.603907] [UCX-connection #1 10.210.0.125:1337] released
[1607977457.603914] [DEMO] Client exit with status 'OK'
[1607977457.603916] [UCX] destroy_connections
[1607977457.604121] [UCX] destroy_connections
[1607977457.604131] [UCX] iomsg receive request 0x2322a78 failed: Request canceled
[1607977457.621452] [jazz30:145487:0]      ptr_map.inl:50   UCX  WARN  ptr map 0x2304160 contains 40045 elements on destroy
```
So, I guess the PR fixes the original problem.

**Comment by dmitrygx** [comment]
failure is #6027

**Comment by alinask** [comment]
> > Could you please run io_demo with this PR with an Eth IP (not RoCE)? to make sure there isn't a leak.
> 
> @alinask sure, thank you for giving me a reproducer.
> I ran the test as you requested. It works fine when I specified IP over commodity Ethernet, no leaks were revealed during the test.
> I tested it _with_ and _without_ `UCX_TLS="tcp,sockcm"`, because the master branch selects IB (if it exists) even if we specified an IP address of an Ethernet device (establishes a connection thru sockcm and TCP, and then switches to more optimal transport).
> 
> ```
> [1607977721.761084] [UCX-connection #1 10.210.0.125:1337] destroying, ep is 0x7fec8111d000
> [1607977721.761133] [UCX-connection #1 10.210.0.125:1337] closing ep 0x7fec8111d000 mode force
> [1607977721.761254] [UCX-connection #1 10.210.0.125:1337] released
> [1607977721.761260] [DEMO] Client exit with status 'OK'
> [1607977721.761263] [UCX] destroy_connections
> [1607977721.761505] [UCX] destroy_connections
> [1607977721.761519] [UCX] iomsg receive request 0xa2ba80 failed: Request canceled
> ```
> 
> Running the test on the master branch completes with the warning at the end of the testing when `UCX_TLS="tcp,sockcm"` is specified (but if an IB transport is used, it completes w/o any warnings):
> 
> ```
> ...
> [1607977457.603803] [UCX-connection #1 10.210.0.125:1337] destroying, ep is 0x7f8aa1787000
> [1607977457.603821] [UCX-connection #1 10.210.0.125:1337] closing ep 0x7f8aa1787000 mode force
> [1607977457.603907] [UCX-connection #1 10.210.0.125:1337] released
> [1607977457.603914] [DEMO] Client exit with status 'OK'
> [1607977457.603916] [UCX] destroy_connections
> [1607977457.604121] [UCX] destroy_connections
> [1607977457.604131] [UCX] iomsg receive request 0x2322a78 failed: Request canceled
> [1607977457.621452] [jazz30:145487:0]      ptr_map.inl:50   UCX  WARN  ptr map 0x2304160 contains 40045 elements on destroy
> ```
> 
> So, I guess the PR fixes the original problem.

@dmitrygx thanks for testing this!

**Review by alinask — APPROVED** [review]

**Comment by dmitrygx** [comment]
@yosefe @brminich @alinask ok to merge?

**Pull request merged**